rust-sdk/core/src/quote/swap.rs

```diff
@@ -0,0 +1,677 @@
+use core::cmp::{max, min};
+
+use crate::{
+    adjust_amount, get_amount_delta_a, get_amount_delta_b, get_next_sqrt_price_from_a,
+    get_next_sqrt_price_from_b, inverse_adjust_amount, sqrt_price_to_tick_index,
+    tick_index_to_sqrt_price, AdjustmentType, ExactInSwapQuote, ExactOutSwapQuote, TickArrayFacade,
+    TickArraySequence, TickFacade, TransferFee, WhirlpoolFacade, U128,
+};
+
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::*;
+
+const MIN_SQRT_PRICE: u128 = 4295048016;
+const MAX_SQRT_PRICE: u128 = 79226673515401279992447579055;
+
+/// Computes the exact input or output amount for a swap transaction.
+///
+/// # Arguments
+/// - `token_in`: The input token amount.
+/// - `specified_token_a`: If `true`, the input token is token A. Otherwise, it is token B.
+/// - `slippage_tolerance`: The slippage tolerance in basis points.
+/// - `whirlpool`: The whirlpool state.
+/// - `tick_array_0`: The tick array at the current tick index.
+/// - `tick_array_plus_1`: The tick array at the current tick offset plus 1.
+/// - `tick_array_plus_2`: The tick array at the current tick offset plus 2.
+/// - `tick_array_minus_1`: The tick array at the current tick offset minus 1.
+/// - `tick_array_minus_2`: The tick array at the current tick offset minus 2.
+/// - `transfer_fee_a`: The transfer fee for token A.
+/// - `transfer_fee_b`: The transfer fee for token B.
+///
+/// # Returns
+/// The exact input or output amount for the swap transaction.
+#[allow(clippy::too_many_arguments)]
+#[cfg_attr(feature = "wasm", wasm_bindgen(js_name = swapQuoteByInputToken, skip_jsdoc))]
+pub fn swap_quote_by_input_token(
+    token_in: U128,
+    specified_token_a: bool,
+    slippage_tolerance: u16,
+    whirlpool: WhirlpoolFacade,
+    tick_array_0: TickArrayFacade,
+    tick_array_plus_1: TickArrayFacade,
+    tick_array_plus_2: TickArrayFacade,
+    tick_array_minus_1: TickArrayFacade,
+    tick_array_minus_2: TickArrayFacade,
+    transfer_fee_a: Option<TransferFee>,
```

Comment:
Yeah this is another one of those `no_std` quirks. Vec is part of std so ideally we'd not use it here. We can then only do an array since wasm cannot handle borrowed values (slice). I then just figured why not give all 5 a param name instead of [TickArrayFacade;5]

Reply:
If we can use Option (it is used for transfer_fee_a), Option will help us.
If we can use Enum with wasm, we can implement our enum for Option<TickArrayFacade>.
```
enum OptionalTickArray {
   None
   Some(TickArrayFacade)
```